### PR TITLE
usdc_reader benchmark

### PR DESCRIPTION
**Pull request by 0xnogo**

### CODE REVIEW

File: core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go [hunk 1]
```diff
@@ -201,7 +207,147 @@ func Test_USDCReader_MessageHashes(t *testing.T) {
...
+	var logs []logpoller.Log
+	messageSentEventSig := abi.Events["MessageSent"].ID
+	require.NoError(b, err)
+	messageTransmitterAddress := testEnv.contractAddr
+
+	for i := 0; i < numOfMessages; i++ {
+		nonce := startNonce + uint64(i)
+
+		// Pack the message following exact CCTP format
+		var buf []byte
```
Comment (mateusz-sekara) [inline core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go]: I suggest extracting that to something like `createMessageSentLogPollerData`

File: core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go [hunk 2]
```diff
@@ -201,7 +207,147 @@ func Test_USDCReader_MessageHashes(t *testing.T) {
 		})
 	}
 }
+func Benchmark_MessageHashes(b *testing.B) {
```
Comment (mateusz-sekara) [inline core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go:220]: Should we add the gobench results above the test as a comment? (as a reference)

File: core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go [hunk 3]
```diff
@@ -201,7 +207,147 @@ func Test_USDCReader_MessageHashes(t *testing.T) {
...
+	// Adding a new parameter: tokenCount
+	testCases := []struct {
+		name       string
+		msgCount   int
+		startNonce uint64
+		tokenCount int
+	}{
+		{"Small_Dataset", 100, 1, 5},
+		{"Medium_Dataset", 10_000, 1, 10},
+		{"Large_Dataset", 100_000, 1, 50},
```
Comment (mateusz-sekara) [inline core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go:232]: Nice, was there any major difference between these 3 datasets?

File: core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go (hunk 2 above)
Comment (mateusz-sekara) [reply core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go:220]: Ideally we want to track the performance regressions somehow (maybe it's not in the scope for 1st iteration, but wanted to call it out anyway). Some time ago, I saw this tool https://pkg.go.dev/golang.org/x/perf/cmd/benchstat, but not sure how much it fits our use case

File: core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go [hunk 4]
```diff
@@ -201,7 +207,147 @@ func Test_USDCReader_MessageHashes(t *testing.T) {
...
+		data = append(data, message[:]...)
+
+		// Create topics array with just the event signature
+		topics := [][]byte{
+			messageSentEventSig[:], // Topic[0] is event signature
+		}
+
+		// Create log entry
+		logs = append(logs, logpoller.Log{
+			EvmChainId:     ubig.New(big.NewInt(int64(uint64(source)))),
```
Comment (mateusz-sekara) [inline core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go]: hmm, is that right? We should use chain_id, not the chain selectors which is stored in `source`, right?

**Review by mateusz-sekara — COMMENTED** [review]
Looks good!

File: core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go (hunk 2 above)
Comment (0xnogo) [reply core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go:220]: I agree, benchmark tests provide very few impact if not compared. We could add a tool like benchstat in the CI and fail if we noticed a reg.

File: core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go (hunk 3 above)
Comment (0xnogo) [reply core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go:232]:
```
// Benchmark_MessageHashes/Small_Dataset-14       3723        272421 ns/op       126949 B/op      2508 allocs/op
// Benchmark_MessageHashes/Medium_Dataset-14       196       6164706 ns/op      1501435 B/op     20274 allocs/op
// Benchmark_MessageHashes/Large_Dataset-14          7     163930268 ns/op     37193160 B/op    463954 allocs/op
```

It's going linearly. I think that the remappings we have (first the messageSentEvents and then out) are a big offender (apart for the GetQuery).

File: core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go (hunk 4 above)
Comment (0xnogo) [reply core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go]: The simulated backend is created with the source chain. The only part we are using the `ChainID` is for the transactor creation (as we don't have any other choice here, geth is complaining about it).

**Review by mateusz-sekara — APPROVED** [review]

**Review by asoliman92 — APPROVED** [review]

**Pull request merged**